src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -320,6 +324,33 @@ public void ConnectHeaders_SupplyHeadersThenSupplyMore_Both_Tip_PeerId_Maps_Shou
             Assert.True(peerTipsByPeerIdBefore.Values.Count == peerTipsByPeerIdAfter.Values.Count);
         }
 
+        /// <summary>
+        /// Issue 4 @ Create chained header tree component #1321
+        /// Supply headers where half of them are new and half are old. 
+        /// Make sure that ChainedHeader was created for new ones.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_HalfOldHalfNew_ShouldCreateHeadersForNew()
+        {
+            const int initialChainSize = 20, chainExtensionSize = 20;
+
+            // Initialize tree with h1->h20.
+            TestContext testContext = new TestContextBuilder().WithInitialChain(initialChainSize).Build();
+            ChainedHeaderTree chainedHeaderTree = testContext.ChainedHeaderTree;
+            ChainedHeader chainTip = testContext.InitialChainTip;
+
+            // Extend chain from h21->h40.
+            ChainedHeader newChainTip = testContext.ExtendAChain(chainExtensionSize, chainTip);
+            List<BlockHeader> listOfOldAndNewHeaders = testContext.ChainedHeaderToList(newChainTip, initialChainSize + chainExtensionSize);
+            
+            // Supply both old and new headers. 
+            chainedHeaderTree.ConnectNewHeaders(1, listOfOldAndNewHeaders);
+
+            // ChainedHeader tree entries are created for all new BlockHeaders.
+            IEnumerable<uint256> hashesOfNewBlocks = listOfOldAndNewHeaders.Select(x => x.GetHash()).TakeLast(chainExtensionSize);
+            Assert.True(hashesOfNewBlocks.All(x => chainedHeaderTree.GetChainedHeadersByHash().Keys.Contains(x)));
```

Comment:
One of those Should contain type asserts might be better here otherwise an error you will get here on failure will just be "expected false, got true)

This is not just a "style" opinion, the goal is a more specific error on failure.